### ISSUE TITLE
Using fixed-size Eigen types

### ISSUE THREAD

**Issue by andrjohns** [stan-dev/stanc3#666]
## Description
It's generally mentioned in the Eigen doc that using fixed-size Eigen types for smaller objects will generally be more performant (mentioned [here](https://eigen.tuxfamily.org/dox/group__TutorialMatrixClass.html#title8) and [here](http://tudat.tudelft.nl/tutorials/basics/externalLibraries/eigen.html#fixed-size-vs-dynamic-size)), since the fixed-size matrix gets allocated on the stack and can be explicitly unrolled.

Where the dynamically-sized matrix would be declared as:
```
Eigen::Matrix<double, -1, -1> mat(4,4);
```
The fixed-size matrix is declared as:
```
Eigen::Matrix<double, 4, 4> mat;
```

However, given the templated nature of Stan, it's hard to know in advance whether a matrix should be instantiated as fixed- or dynamic-size. To get around this, we could create a new typedef that instantiates to fixed or dynamic, depending on the dimension size:
```
template <int D>
struct dim {
    enum { value = (D > 4 ? Eigen::Dynamic : D) };
};

template <typename T, int R, int C>
using stan_matrix = Eigen::Matrix<T, dim<R>::value, dim<C>::value>;
```

But I might be misunderstanding things here. @t4c1 does this make sense?

#### Current Version:
v3.3.0

**Comment by andrjohns** [comment]
I ran some benchmarks comparing the performance of fixed and dynamic matrices of size 4x4, 8x8, and 16x16 with vars and doubles. 

```exp```:
```
----------------------------------------------------------------
Benchmark                      Time             CPU   Iterations
----------------------------------------------------------------
Fixed_4x4_exp              0.000 ns        0.000 ns   1000000000
Dynamic_4x4_exp             51.5 ns         51.5 ns     13647890
Fixed_8x8_exp              0.230 ns        0.230 ns   1000000000
Dynamic_8x8_exp              160 ns          160 ns      4380417
Fixed_16x16_exp            0.230 ns        0.230 ns   1000000000
Dynamic_16x16_exp            597 ns          597 ns      1182666
Fixed_4x4_var_exp            293 ns          293 ns      2148349
Dynamic_4x4_var_exp          399 ns          399 ns      2404058
Fixed_8x8_var_exp           1066 ns         1065 ns       634387
Dynamic_8x8_var_exp         1032 ns         1032 ns       676960
Fixed_16x16_var_exp         7236 ns         7235 ns       171191
Dynamic_16x16_var_exp       4051 ns         4051 ns       172552
```
Looks like there are dramatic speed-ups with ```double``` types, and only marginal differences with ```var``` until 16x16 when Dynamic outperforms

Matrix multiplication:
```
-------------------------------------------------------------------
Benchmark                         Time             CPU   Iterations
-------------------------------------------------------------------
Fixed_4x4_matmul              0.000 ns        0.000 ns   1000000000
Dynamic_4x4_matmul             58.1 ns         58.1 ns     12157907
Fixed_8x8_matmul                119 ns          119 ns      5833427
Dynamic_8x8_matmul              182 ns          182 ns      3902449
Fixed_16x16_matmul              523 ns          523 ns      1342618
Dynamic_16x16_matmul            598 ns          598 ns      1171889
Fixed_4x4_var_matmul            448 ns          448 ns      1494162
Dynamic_4x4_var_matmul          565 ns          565 ns      1537562
Fixed_8x8_var_matmul           1515 ns         1515 ns       461309
Dynamic_8x8_var_matmul         2191 ns         2191 ns       452395
Fixed_16x16_var_matmul         5697 ns         5696 ns       122574
Dynamic_16x16_var_matmul       5810 ns         5810 ns       120619
```
There are speed-ups for both ```double``` and ```var``` types, with diminishing returns  as the dimenstions get larger

**Comment by bob-carpenter** [comment]
The discussion on not being able to do make this decision statically in pure C++ is why this isn't already in the compiler.

> Would you mind if I transfer this issue to stanc3?

As @t4c1 notes, the math library needs to be enhanced to handle all these fixed-size matrices everywhere before the compiler could be updated.  For example, we'd want the multiplication of two 3 x 3 fixed-size matrices would return a 3 x 3 fixed-size matrix.

**Comment by rok-cesnovar** [comment]
I am guessing the on-going generalization would help with this?

**Comment by andrjohns** [comment]
> I am guessing the on-going generalization would help with this?

Yeah, all the functions that are generalised to take any Eigen inputs and also with an ```auto``` return type will already work with this

**Comment by rok-cesnovar** [comment]
> As @t4c1 notes, the math library needs to be enhanced to handle all these fixed-size matrices everywhere before the compiler could be updated.

I am fine if we want to transfer it back. My thinking for the transfer to stanc3 was: 
- in math we need to check if functions support it. If they do, then there is no issue in Math. If they don't, then this issue has to wait for generalization, in which case we just mention the generalization Math issue in here as the blocker
- if we decided to do this, the majority of work to close this issue will be in stanc3

@andrjohns the performance results are fantastic!

**Comment by bbbales2** [comment]
One place where this would be handy is in the ODEs. For instance, here: https://github.com/stan-dev/example-models/blob/master/knitr/convert-odes/sir.stan

The malloc cost is significant here (the function itself is very simple). If you run that model and look at the profiler you'll see lots of allocation stuff at the top.

**Comment by rok-cesnovar** [comment]
We first need to generalize the Eigen args in the variadic ode functions. It would be interesting to manually modify a model to see the gain then.

**Comment by bbbales2** [comment]
It might take some work to prototype this up. Check here: https://github.com/stan-dev/math/blob/develop/stan/math/rev/functor/coupled_ode_system.hpp#L123

I think somehow the ODE solver would need to know the state was going to be treated like a compile-time-known-length vector. Presumably that means it'd do types based on the initial conditions, so the initial conditions would need to be recognized as a compile-time known thing.

The allocate below it could be disposed of though with some auto magic.

**Comment by andrjohns** [comment]
@rok-cesnovar Would you mind transferring this back to Math? This needs to be plumbed into the Math library before adding the Stanc3 stuff. Thanks!
